### PR TITLE
auth type: ServicePrincipal

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/ServicePrincipalLoginHelper.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+        // 2. check maven configuration: client, tenant
+        if (StringUtils.isNoneBlank(configuration.getClient(), configuration.getTenant())) {
+            configuration.validate();
+            return mavenSettingLogin(AuthMethod.MAVEN_CONFIGURATION, configuration);
+        }
+        return null;
+    }
+
+    private static AzureCredentialWrapper mavenSettingLogin(AuthMethod method, MavenAuthConfiguration configuration) {
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/ServicePrincipalLoginHelper.java:40]: We may use different method for `MAVEN_SETTINGS` and `MAVEN_CONFIGURATION` auth as there are few common codes to share.

Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/ServicePrincipalLoginHelper.java:40]: Like use `getCredentialFromMavenSettings` and `getCredentrialFromMavenConfiguration`, and create `AzureCredentialWrapper` based on credential in `login`

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        configurationFromServer.setServerId(serverId);
+
+        // validate configuration
+        return configurationFromServer.validateAndReturn();
+    }
+
+    /**
+     * Get string value from server configuration section in settings.xml.
+     *
+     * @param configuration the <i>Xpp3Dom</i> object representing the configuration in &gt;server&gt;'.
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:71]: Do we need to encode the `<` here

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java [hunk 3]
```diff
@@ -22,4 +28,38 @@
...
+            messagePrefix = "Cannot find 'tenant'";
+        } else if (StringUtils.isBlank(client)) {
+            messagePrefix = "Cannot find 'client'";
+        } else if (StringUtils.isAllBlank(key, certificate)) {
+            messagePrefix = "Cannot find either 'key' or 'certificate'";
+        } else if (StringUtils.isNotBlank(key) && StringUtils.isNotBlank(certificate)) {
+            messagePrefix = "It is wrong to specify both 'key' and 'certificate'";
+        } else if (StringUtils.isNotBlank(environment) && !AuthHelper.validateEnvironment(environment)) {
+            messagePrefix = "Invalid environment string '%s'";
+        } else if (StringUtils.isNotBlank(httpProxyHost) && StringUtils.isBlank(httpProxyPort)) {
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java:44]: check `httpProxyPort` is not empty but `httpProxyHost` is blank?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java [hunk 4]
```diff
@@ -22,4 +28,38 @@
...
+            messagePrefix = "if you want to use proxy, 'httpProxyHost' and 'httpProxyPort' must both be set";
+        } else if (StringUtils.isNotBlank(httpProxyPort)) {
+            if (!StringUtils.isNumeric(httpProxyPort)) {
+                messagePrefix = String.format("Invalid integer number for httpProxyPort: '%s'", httpProxyPort);
+            } else if (NumberUtils.toInt(httpProxyPort) <= 0 || NumberUtils.toInt(httpProxyPort) > 65535) {
+                messagePrefix = String.format("Invalid range of httpProxyPort: '%s', it should be a number between %d and %d", httpProxyPort, 1, 65535);
+            }
+        }
+
+        final String messagePostfix = StringUtils.isNotBlank(serverId) ? (" in <server>: \"" + serverId + "\" at maven settings.xml.")
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java:54]: Use `String.format("in <server>: \"%s\"  at maven settings.xml.")`?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/ServicePrincipalLoginHelper.java (hunk 1 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/ServicePrincipalLoginHelper.java]: No, they are related to one key: serverId

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+     * @return String value if property exists; otherwise, return null.
+     */
+    private static String getConfiguration(final Xpp3Dom configuration, final String property) {
+        final Xpp3Dom node = configuration.getChild(property);
+        return Objects.isNull(node) ? null : node.getValue();
+    }
+
+    private static boolean isValueEncrypted(String value) {
+        return value != null && value.startsWith("{") && value.endsWith("}");
+    }
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:82]: StringUtils.startWith.../endWith... ?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java [hunk 6]
```diff
@@ -22,4 +28,39 @@
...
+                messagePrefix = String.format("Invalid range of httpProxyPort: '%s', it should be a number between %d and %d", httpProxyPort, 1, 65535);
+            }
+        }
+
+        final String messagePostfix = StringUtils.isNotBlank(serverId) ? (" in server: '" + serverId + "' at maven settings.xml.")
+                : " in <auth> configuration.";
+        if (Objects.nonNull(messagePrefix)) {
+            throw new InvalidConfigurationException(messagePrefix + messagePostfix);
+        }
+    }
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java:60]: how about to use Preconditions.checkArgument(expression, message) here ?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    private static String decryptMavenProtectedValue(SettingsDecrypter settingsDecrypter, String propertyName, String value) throws MavenDecryptException {
+        final Server server = new Server();
+        server.setPassword(value);
+        final SettingsDecryptionRequest request = new DefaultSettingsDecryptionRequest(server);
+        final SettingsDecryptionResult result = settingsDecrypter.decrypt(request);
+        for (final SettingsProblem problem : result.getProblems()) {
+            if (problem.getSeverity() == SettingsProblem.Severity.ERROR || problem.getSeverity() == SettingsProblem.Severity.FATAL) {
+                // for java 8+, it is ok to use operator '+' for string concatenation
+                throw new MavenDecryptException(propertyName, value, "Unable to decrypt value(" + value + ") from settings.xml: " + problem);
```
Comment (shenqianjin) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:92]: I prefer to user SettingsProblem.Severity.ERROR.equals(problem.getSeverity()) ...

Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:92]: enum prefers ==

Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:92]: https://stackoverflow.com/questions/1750435/comparing-java-enum-members-or-equals

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java (hunk 6 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java:60]: checkArgument will throw IllegalArgumentException only, we want to throw InvalidConfigurationException

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java (hunk 5 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:82]: No difference in reading, the logic in StringUtils is more complex
image

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+import org.apache.maven.settings.building.SettingsProblem;
+import org.apache.maven.settings.crypto.DefaultSettingsDecryptionRequest;
+import org.apache.maven.settings.crypto.SettingsDecrypter;
+import org.apache.maven.settings.crypto.SettingsDecryptionRequest;
+import org.apache.maven.settings.crypto.SettingsDecryptionResult;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+
+import java.util.Objects;
+
+public class MavenSettingHelper {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:25]: is there any difference between `Utils` and `Helper`, I can see both in code project. I suggest only use `Utils`

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java [hunk 9]
```diff
@@ -22,4 +28,39 @@
     private String serverId;
     private String httpProxyHost;
     private String httpProxyPort;
+
+    public void validate() throws InvalidConfigurationException {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java:32]: the responsibility of validation should not be assigned to POJO itself. 
i.e. `validate` method should be moved out to some other place.

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/tools/ServicePrincipalLoginHelper.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+import com.microsoft.azure.tools.auth.maven.MavenSettingHelper;
+import com.microsoft.azure.tools.auth.model.AuthMethod;
+import com.microsoft.azure.tools.auth.model.AzureCredentialWrapper;
+import com.microsoft.azure.tools.auth.model.MavenAuthConfiguration;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.settings.crypto.SettingsDecrypter;
+
+public class ServicePrincipalLoginHelper {
+    public static AzureCredentialWrapper login(MavenAuthConfiguration configuration, MavenSession session, SettingsDecrypter settingsDecrypter)
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/tools/ServicePrincipalLoginHelper.java:25]: it's some weird that I have to provide `SettingsDecrypter` when I login.

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java [hunk 11]
```diff
@@ -22,4 +28,39 @@
...
+        String messagePrefix = null;
+        if (StringUtils.isBlank(tenant)) {
+            messagePrefix = "Cannot find 'tenant'";
+        } else if (StringUtils.isBlank(client)) {
+            messagePrefix = "Cannot find 'client'";
+        } else if (StringUtils.isAllBlank(key, certificate)) {
+            messagePrefix = "Cannot find either 'key' or 'certificate'";
+        } else if (StringUtils.isNotBlank(key) && StringUtils.isNotBlank(certificate)) {
+            messagePrefix = "It is wrong to specify both 'key' and 'certificate'";
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java:41]: it seems this message is wrong.
and seems the upper 2 `if`s can be merged into one
```
if(StringUtils.isAllBlank(key, certificate) || StringUtils.isNoneBlank(key, certificate)){
    messagePrefix = "You have to specify either `key` or `certificate`";
}
```

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /**
+     * Get string value from server configuration section in settings.xml.
+     *
+     * @param configuration the <code>Xpp3Dom</code> object representing the configuration in &lt;server&gt;.
+     * @param property           the property name
+     * @return String value if property exists; otherwise, return null.
+     */
+    private static String getConfiguration(final Xpp3Dom configuration, final String property) {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:75]: rename to `getValue`/`getPropertyValue`.
`Configuration` should be `MavenAuthConfiguration` in current scope.

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        if (server == null) {
+            throw new InvalidConfigurationException(String.format("serverId '%s' cannot be found in maven settings.xml.", serverId));
+        }
+        final MavenAuthConfiguration configurationFromServer = new MavenAuthConfiguration();
+        final Xpp3Dom configuration = (Xpp3Dom) server.getConfiguration();
+        configurationFromServer.setServerId(serverId);
+        if (configuration == null) {
+            return configurationFromServer;
+        }
+        configurationFromServer.setTenant(getConfiguration(configuration, "tenant"));
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:50]: method name `getConfiguration` make me confused. 
get a auth configuration from a configuration? and tenant is auth configuration?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+     * @param settingsDecrypter the decrypter
+     * @param serverId          the server id
+     * @return the auth configuration
+     * @throws AzureLoginException when there are errors decrypting maven generated password
+     */
+    public static MavenAuthConfiguration getAuthConfigurationFromServer(MavenSession session, SettingsDecrypter settingsDecrypter, String serverId)
+            throws AzureLoginException {
+        if (StringUtils.isBlank(serverId)) {
+            throw new IllegalArgumentException("Parameter 'serverId' cannot be null or empty.");
+        }
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:39]: this check seems no need.

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+     * Get auth configuration from serverId.
+     *
+     * @param session           the maven session
+     * @param settingsDecrypter the decrypter
+     * @param serverId          the server id
+     * @return the auth configuration
+     * @throws AzureLoginException when there are errors decrypting maven generated password
+     */
+    public static MavenAuthConfiguration getAuthConfigurationFromServer(MavenSession session, SettingsDecrypter settingsDecrypter, String serverId)
+            throws AzureLoginException {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:36]: it seems this method did 2/3 tasks:
get server from session.
(get raw configuration from server)
build configuration from raw configuration.

so why not extract 2 methods:
`Xpp3Dom getServerRawConfigution(MavenSession session, String serverId)`
`MavenAuthConfiguration buildConfiguration(Xpp3Dom rawConfiguration)`

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+        if (StringUtils.isBlank(serverId)) {
+            throw new IllegalArgumentException("Parameter 'serverId' cannot be null or empty.");
+        }
+        final Server server = session.getSettings().getServer(serverId);
+        if (server == null) {
+            throw new InvalidConfigurationException(String.format("serverId '%s' cannot be found in maven settings.xml.", serverId));
+        }
+        final MavenAuthConfiguration configurationFromServer = new MavenAuthConfiguration();
+        final Xpp3Dom configuration = (Xpp3Dom) server.getConfiguration();
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:45]: ⬆️ getRawConfiguration() from session and server.
⬇️ buildConfiguration() from raw configuration

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    private static String getConfiguration(final Xpp3Dom configuration, final String property) {
+        final Xpp3Dom node = configuration.getChild(property);
+        return Objects.isNull(node) ? null : node.getValue();
+    }
+
+    private static boolean isValueEncrypted(String value) {
+        return value != null && value.startsWith("{") && value.endsWith("}");
+    }
+
+    private static String decryptMavenProtectedValue(SettingsDecrypter settingsDecrypter, String propertyName, String value) throws MavenDecryptException {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:84]: it's some weird that the caller have to provide `propertyName` when decrypt a value?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java (hunk 8 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:25]: MavenSettingHelper  is moved from auth-helper, will rename all helper to util in later PR

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java (hunk 15 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/maven/MavenSettingHelper.java:36]: will move this class out of this module in later PR.

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java (hunk 11 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/model/MavenAuthConfiguration.java:41]: if merged the error message contains two instructions, which is not intuitive

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/tools/ServicePrincipalLoginHelper.java (hunk 10 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/tools/ServicePrincipalLoginHelper.java:25]: will refactor it later

**Comment by wangmingliang-ms** [comment]
It's acceptable to merge into feature branch first, but please remember to fix the comments in later PRs.

**Review by wangmingliang-ms — APPROVED** [review]
It's acceptable to merge into feature branch first, but please remember to fix the comments in later PRs into the same feature branch

**Pull request merged**